### ISSUE TITLE
What is the Training Data format used in sklearn_intent_classifier.py

### ISSUE THREAD

**Issue by sabbiu** [RasaHQ/rasa_nlu#472]
As in Line 94 of  [https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/classifiers/sklearn_intent_classifier.py#L94](https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/classifiers/sklearn_intent_classifier.py#L94), there is a function called get('text_features'). 

What does this 'text_features' refer to? 
Provide me the training data set that fits this classifier.

**Comment by wrathagom** [comment]
The featurizers are the ones to provide `text_features` https://github.com/RasaHQ/rasa_nlu/tree/master/rasa_nlu/featurizers.

All of these are part of the pipeline that Rasa executes: https://rasa-nlu.readthedocs.io/en/latest/pipeline.html

The training data format going into the pipeline is well documented here https://rasa-nlu.readthedocs.io/en/latest/dataformat.html

Please let me know if that helps or not.

**Comment by sabbiu** [comment]
Thanks. It helped me to better understand the process. But, still I am having some issues. Correct me if I am wrong.

**Understanding**
As per [regex_featurizer.py](https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/featurizers/regex_featurizer.py),
If I am not wrong, the `text_features` contain an array of 1 and 0 for respective `known_patterns` as `'text_features': [1,0,0,1,0]` and token is assigned with pattern index if present/matched as `'token': { 'pattern': 2 }`

Is the above same with [mitie_featurizer.py](https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/featurizers/mitie_featurizer.py) ? 
`# Because of having to look into further details about mitie, I haven't gone through this file`
`# If not, provide me with the format text_features`

Now with the file [sklearn_intent_classifier.py](https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/classifiers/sklearn_intent_classifier.py)
`x is the input for the classification with text_features as [1,0,0,1,0] for all the training_examples`
`Y is the output with array of respective intent converted to index for all the training_examples`

**Issues**
Some queries might contain same kind of text_features but belong to different intent. Please correct me if I am missing some point.

Also, Few Queries:
By this, how does it cover the training_examples where some of them may have text_features/entity and some of them may not have any.
Example might include `greeting` intent (might not have text_features) and `asking_for_location` intent 
(will have text_features) both.

**In short**
What are the input and output for training the model? How does it cover training_example (some with text_features and some without)? What does the `text_features` actually contain after running `regex_featurizer.py` and `mitie_featurizer.py`?

**Comment by wrathagom** [comment]
Reading Alan's blog on DIY for Bot developers may be the best place for you to start to get answers to your questions https://medium.com/rasa-blog/do-it-yourself-nlp-for-bot-developers-2e2da2817f3d.

I'm going to close this for now, but please feel free to post back if you've got more questions or join the discussion on [Gitter](https://gitter.im/RasaHQ/rasa_nlu?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge).

**Comment by sabbiu** [comment]
I learnt that the words are tokenized and entities are identified. Now, how are these information actually used for training? The word that I'm using might not be in the word2vec model.

Also, if you have the intent data in which I can train and test the classifier upon. Please, provide it to me. Thanks.

**Comment by tmbo** [comment]
@sabbiu you are completely right there, a word might not be part of the vocabulary the vector model is trained to recognize. In that case (and that is common for very specific domains) there are two solutions to approach this:
- add the `featurizer_ngram` to the processing pipeline (which adds features to every token and allows the classifier to use the words for the classification (intent) even if they are not part of the vocabulary)
- fine tune / retrain a specific vector model tailored for your domain (way more work but might be a good idea if there are a lot of very important words in your domain that are not part of the vector model).

At the moment we don't provide any intent data sets (there is a small example dataset in `data/examples/rasa/demo-rasa.json` but that is really just for showcasing the format of the training data). This might change in the future, though.

To might things a bit easier, please try not to comment on closed issues (we might miss your comment and forget to address your question). Please just open a new issue.
